src/bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,391 @@
+            is_directory=False,
+            parameter="data_uris",
+            num=len(self.assets) + 1,
+        )
+        self.assets.append(new_asset)
+        self.data_uris.append(new_datum_uri)
+
+        super().consume_stream_datum(doc)
+
+    @property
+    def adapter_parameters(self) -> Dict:
+        return {"data_uris": self.data_uris}
+
+
+STREAM_HANDLER_REGISTRY = {
+    "application/x-hdf5": HDF5StreamHandler,
+    "multipart/related;type=image/tiff": TIFFStreamHandler,
+}
+
+
+class TiledWriter:
+    "Write metadata and data from Bluesky documents into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self._run_router = RunRouter([self._factory])
+
+    def _factory(self, name, doc):
+        return [_RunWriter(self.client)], []
+
+    @classmethod
+    def from_uri(cls, uri, **kwargs):
+        client = from_uri(uri, **kwargs)
+        return cls(client)
+
+    @classmethod
+    def from_profile(cls, profile, **kwargs):
+        client = from_profile(profile, **kwargs)
+        return cls(client)
+
+    def __call__(self, name, doc):
+        self._run_router(name, doc)
+
+
+class _RunWriter(DocumentRouter):
+    "Write the document from one Bluesky Run into Tiled."
+
+    def __init__(self, client: BaseClient):
+        self.client = client
+        self.root_node: Union[None, Container] = None
+        self._desc_nodes: Dict[str, Container] = {}  # references to descriptor containers by their uid's
+        self._sres_nodes: Dict[str, BaseClient] = {}
+        self._sres_cache: Dict[str, StreamResource] = {}
+        self._handlers: Dict[str, StreamHandlerBase] = {}
+
+    def _ensure_sres_backcompat(self, sres: StreamResource) -> StreamResource:
+        """Kept for back-compatibility with old StreamResource schema from event_model<1.20.0
+
+        Will make changes to and return the _same_instance_ of StreamRsource dictionary.
+        """
+
+        if ("mimetype" not in sres.keys()) and ("spec" not in sres.keys()):
+            raise RuntimeError("StreamResource document is missing a mimetype or spec")
+        else:
+            sres["mimetype"] = sres.get("mimetype", MIMETYPE_LOOKUP[sres["spec"]])
+        if "parameters" not in sres.keys():
+            sres["parameters"] = sres.pop("resource_kwargs", {})
+        if "uri" not in sres.keys():
+            file_path = sres.pop("root").strip("/") + "/" + sres.pop("resource_path").strip("/")
+            sres["uri"] = "file://localhost/" + file_path
+
+        return sres
+
+    def start(self, doc):
+        self.root_node = self.client.create_container(
+            key=doc["uid"],
+            metadata={"start": doc},
+            specs=[Spec("BlueskyRun", version="1.0")],
+        )
+
+    def stop(self, doc):
+        if self.root_node is None:
+            raise RuntimeError("RunWriter is properly initialized: no Start document has been recorded.")
+
+        doc = dict(doc)
+        doc.pop("start", None)
```

Comment:
It's not a requirement. Removing it would serve "one source of truth" but I think that there is a case to be made for retaining the complete document in its original form here, to be make it easy to reconstruct the documents correctly, and also for back-compat.

Reply:
Yes, the main reason to pop the uid of the Start document from the stored Stop was to eliminate redundancy, but I also completely agree with the argument in favour of maintaining consistency between StartDocument and StopDocument.

On the other hand, what I personally find somewhat confusing is that same keys refer to different types of values depending on the dictionary/structure. E.g., in the metadata of Tiled's `root_node`, `start` and `stop` keys refer to the corresponding actual documents, but in the StopDocument itself, `start` reference the _id_ of the StartDocument, and by extension -- the entire BlueskyRun. Similarly `descriptor` key in StreamResources really contains `descriptor_id`, and so on. One way to solve this would be by calling the fields in Tiled metadata `start_document`and `stop_document`, but this seems to be overly verbose and wouldn't work for descriptors, where the entire metadata dictionary of the same Tiled node depends on multiple descriptor documents, in general.

I have reinstated the `start` field in the saved StopDocuments to keep it in its original form, as suggested, but this part likely needs a closer look and warrants a further discussion in line with Dan's [comment](https://github.com/bluesky/bluesky/pull/1660#discussion_r1587763803) below. My preference would be to keep as much information from the Bluesky documents as possible, but to parse them before storing in the Tiled database and remove all duplications. I agree that reconstructing the original documents would require some work, but it wouldn't be very difficult.